feat: zksolc link developer doc (#71)
matter-labs/foundry-zksync-book: src/zksync-specifics/developer-guide/linking.md

#  Deploy-time linking with zksolc

> Contrary to `solc`, unlinked bytecode in `zksolc` contain their link references using the ELF format, thus a simple text search & replace is not sufficient.

Reviewer: ```suggestion
Contrary to `solc`, unlinked bytecode in `zksolc` contains their link references using the ELF format, thus a simple text search & replace is not sufficient.
```

The main difference between the two is that tests collect all missing library references across all contracts and collate them together to then deploy them for each test contract, effectively deploying all libraries for all tests; contrary to scripts where `foundry` will collect references and link only libraries in the target script's dependency tree.

Therefore, the general flow is:
1. collect all dependencies of the target(s)
2. perform linking by precalculating the address of each library
  a. using `CREATE` with a single pass (tests only allow this methodology)
  b. using `CREATE2` with multiple passes until the list exhaustion
3. collect the linked contracts as well as the libraries to deploy
  a. specifically for zksync: register linked contracts in `DualCompiledContracts`
4. perform deployment according to the selected opcode
5. deploy test/script & invoke 

We will now go into the details of each step, explaining the differences between the EVM flow and EraVM's.

### Dependency resolution

> The main difference with dependency resolution is with the introduction of factory deps, this means that factory deps may remain unlinked during compilation due to their use of a (missing) library. 

Reviewer: ```suggestion
The main difference in dependency resolution is the introduction of factory dependencies. This means that factory dependencies may remain unlinked during compilation due to their reliance on a (missing) library.
```
